File: app/src/main/java/com/elitec/alejotaller/feature/notifications/data/mappers/Mappers.kt
package com.elitec.alejotaller.feature.notifications.data.mappers

import com.elitec.alejotaller.feature.notifications.data.dto.PromotionDto
import com.elitec.alejotaller.feature.notifications.data.models.PromotionEvent
import com.elitec.shared.core.feature.notifications.domain.entity.Promotion

fun PromotionDto.toDomain(): Promotion = Promotion(
    id = id,
    productId = productId,
    title = title,
    message = message,
    imageUrl = imageUrl,
    oldPrice = oldPrice,
    currentPrice = currentPrice,
    validFromEpochMillis = validFromEpochMillis,
    validUntilEpochMillis = validUntilEpochMillis,
    kind = kind,
    status = status,
    source = source,
)

fun Promotion.toDto(): PromotionDto = PromotionDto(
    id = id,
    productId = productId,
    title = title,
    message = message,
    imageUrl = imageUrl,
    oldPrice = oldPrice,
    currentPrice = currentPrice,
    validFromEpochMillis = validFromEpochMillis,
    validUntilEpochMillis = validUntilEpochMillis,
    kind = kind,
    status = status,
    source = source,
)

private const val DEFAULT_PROMO_TTL = 1000L * 60L * 60L * 24L * 7L

fun PromotionEvent.toDomainPromotion(): Promotion {
    val now = System.currentTimeMillis()

    return Promotion(
        id = id.ifBlank { "promo-$now" },
        productId = null,
        title = title,
        message = message,
        imageUrl = imageUrl,
        oldPrice = oldPrice,
        currentPrice = currentPrice,
        validFromEpochMillis = validFromEpochMillis ?: now,
        validUntilEpochMillis = validUntilEpochMillis ?: (now + DEFAULT_PROMO_TTL),
    )
}

File: web/src/core/feature/notification/data/mapper/Mappers.ts
import type { PromotionDTO } from "../dto/PromotionDTO"
import type { Promotion, PromotionKind, PromotionStatus } from "../../domain/entity/Promotion"

function asKind(value: unknown): PromotionKind | undefined {
    if (value === "banner" || value === "product_discount") return value
    return undefined
}

function asStatus(value: unknown): PromotionStatus | undefined {
    if (value === "draft" || value === "active" || value === "ended" || value === "cancelled") {
        return value
    }
    return undefined
}

export function promotionFromDTO(dto: PromotionDTO): Promotion {
    return {
        id: dto.$id,
        productId: dto.productId ?? null,
        title: dto.title,
        message: dto.message,
        imageUrl: dto.imageUrl ?? null,
        oldPrice: dto.oldPrice ?? null,
        currentPrice: dto.currentPrice ?? null,
        validFromEpochMillis: dto.validFromEpochMillis,
        validUntilEpochMillis: dto.validUntilEpochMillis,
        source: dto.source === "manual" ? "manual" : dto.source === "automatic" ? "automatic" : undefined,
        kind: asKind(dto.kind),
        status: asStatus(dto.status),
    }
}
